mappings/net/minecraft/entity/vehicle/AbstractMinecartEntity.mapping

```diff
@@ -22,13 +25,24 @@ CLASS net/minecraft/class_1688 net/minecraft/entity/vehicle/AbstractMinecartEnti
 		ARG 7 z
 	METHOD method_18803 willHitBlockAt (Lnet/minecraft/class_2338;)Z
 		ARG 1 pos
+	METHOD method_22864 getDirectionOffsetsForRailShape (Lnet/minecraft/class_2768;)Lcom/mojang/datafixers/util/Pair;
 	METHOD method_7504 getMaxOffRailSpeed ()D
+	METHOD method_7505 snapPositionToRailWithOffset (DDDD)Lnet/minecraft/class_243;
```

Comment:
Just calculates the visual position of the minecart on a rail, not really snapping.
Also is this offset y or x or z offset?

Reply:
It's an offset along the rail. It's unspecified which way is + and which way is -. The renderer calls it with both +0.3 and -0.3 so it doesn't matter which way is which.

It applies an offset along the rail direction and then calls `snapPositionToRail`. The returned position is always snapped.

Reply:
how is the offset used within this method?

Reply:
The offset is how far along the rail's direction vector the coordinates should be adjusted by.

The function is basically: `return snapPositionToRail(xyz + offset*getRailDirection(getBlockAt(xyz)))`

Reply:
That is not an offset! That is a progress from 0 to 1 indicating what percent of the track should the cart be on.

Reply:
It's not 0 to 1; it can be negative. I think this implementation only works for numbers between -0.5 and 0.5.

Reply:
Let me re-check this code to make sure that's actually right...

Reply:
They are used to get the turn progress of the minecart. Think 0.3 is like the distance in which the cart start to show turn before entering a curvature.

Reply:
Using the names in Enigma with Procyon:

* First, integer10/11/12 are the minecart's integer coordinates.
* For some reason if the minecart is on a slope then its position is in the block above the slope, so correct for that.
* If the minecart isn't on a rail, return null.
* Reset the Y coordinate to an integer. I'm not sure this needs to be fully analyzed since the purpose of the function is clear from what it does with the x/z coordinates.
* Get the vector from one end of the rail block to the other and normalize it. This is double18/20.
* Then offset the passed in x/z by double18/20 * offset. i.e. offset it that far along the rail's direction vector. <- this is the only step where the offset is used. Before this, the x/z are still the minecart's x/z. **<- this is the main part of what the function does**
* Adjust the y somehow for slopes
* Snap this position to the nearest valid rail position.

Reply:
This name should be fine.